File: Backend/src/main/java/com/vidyasahay/vidyasahay/service/LoanSchemeService.java
package com.vidyasahay.vidyasahay.service;

import java.util.List;
import java.util.UUID;

import com.vidyasahay.vidyasahay.dto.request.loanScheme.CreateLoanSchemeRequestDTO;
import com.vidyasahay.vidyasahay.dto.request.loanScheme.LoanEligibilityRequestDTO;
import com.vidyasahay.vidyasahay.dto.request.loanScheme.UpdateLoanSchemeRequestDTO;
import com.vidyasahay.vidyasahay.dto.response.loanScheme.LoanSchemeDetailedResponseDTO;
import com.vidyasahay.vidyasahay.dto.response.loanScheme.LoanSchemeSummaryResponseDTO;

public interface LoanSchemeService {	
	List<LoanSchemeSummaryResponseDTO> getAllLoanSchemes();
	LoanSchemeDetailedResponseDTO getLoanSchemeById(UUID loanSchemeId);
	List<LoanSchemeSummaryResponseDTO> getEligibleLoanSchemes(LoanEligibilityRequestDTO request);
	List<LoanSchemeSummaryResponseDTO> getLoanSchemesCreatedByMe();
	UUID createLoanScheme(CreateLoanSchemeRequestDTO request);
	void updateLoanScheme(UUID loanSchemeId, UpdateLoanSchemeRequestDTO request);
}

File: Frontend/src/app/core/services/loan-scheme.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../../environments/env';
import { LoanSchemeSummary } from '../models/summaryResponse.dto';
import { LoanSchemeDetails } from '../models/loan-scheme-detailed.model';

@Injectable({providedIn: 'root'})
export class LoanSchemeService {

  private readonly API_URL = `${environment.apiUrl}/loan-scheme`;

  constructor(private http: HttpClient) { }

  getLoanSchemes(): Observable<LoanSchemeSummary[]> {
    return this.http.get<LoanSchemeSummary[]>(`${this.API_URL}/all`);
  }

  getActiveLoanSchemes(): Observable<LoanSchemeSummary[]> {
    return this.http.get<LoanSchemeSummary[]>(`${this.API_URL}/active`);
  }

  getEligibleLoanSchemes(requiredLoanAmount: number): Observable<LoanSchemeSummary[]> {
    return this.http.post<LoanSchemeSummary[]>(`${this.API_URL}/eligible`, { requiredLoanAmount });
  }

  getLoanSchemesCreatedByMe(): Observable<LoanSchemeSummary[]> {
    return this.http.get<LoanSchemeSummary[]>(`${this.API_URL}/created-by-me`);
  }

  createLoanScheme(request: Record<string, unknown>): Observable<void> {
    return this.http.post<void>(this.API_URL, request);
  }

  updateLoanScheme(loanSchemeId: string, request: Record<string, unknown>): Observable<void> {
    return this.http.patch<void>(`${this.API_URL}/${encodeURIComponent(loanSchemeId)}`, request);
  }

  getLoanSchemeById(loanSchemeId: string): Observable<LoanSchemeDetails> {
    return this.http.get<LoanSchemeDetails>(`${this.API_URL}/${encodeURIComponent(loanSchemeId)}`);
  }
}
